File: Stack/prefix_to_infix.cpp
#include <iostream>
#include <stack>
using namespace std;

string prefix_to_infix(string &s) {
    stack<string> st;
    int n = s.length();

    // traverse from right to left
    for (int i = n - 1; i >= 0; i--) {
        char c = s[i];

        // Operand
        if ((c >= 'A' && c <= 'Z') || 
            (c >= 'a' && c <= 'z') || 
            (c >= '0' && c <= '9')) {
            st.push(string(1, c));
        } 
        // Operator
        else {
            if (st.size() < 2) {
                cout << "Invalid prefix expression\n";
                return "";
            }
            string op1 = st.top(); st.pop(); // first operand
            string op2 = st.top(); st.pop(); // second operand
            string res = "(" + op1 + c + op2 + ")";
            st.push(res);
        }
    }

    if (st.size() != 1) {
        cout << "Invalid prefix expression\n";
        return "";
    }

    return st.top();
}

int main() {
    string prefix = "*+AB-C"; // example: (A+B)*(C-?)
    string infix = prefix_to_infix(prefix);
    cout << "Prefix: " << prefix << endl;
    cout << "Infix: " << infix << endl;

    return 0;
}

File: java/StackQ/prefix_to_infix.java
package StackQ;

import java.util.Stack;

class prefix_to_infix{

    /*
    ------------------------------------------------------------
    PROBLEM STATEMENT:
    ------------------------------------------------------------
    Given a prefix expression, convert it into its equivalent
    infix expression.

    Example:
    Prefix : *+AB-C
    Infix  : (A+B)*(C-?)
    */

    public static String prefixToInfix(String s) {
        Stack<String> st = new Stack<>();
        int n = s.length();

        // Traverse the prefix expression from RIGHT to LEFT
        for (int i = n - 1; i >= 0; i--) {
            char c = s.charAt(i);

            // If operand, push to stack
            if (Character.isLetterOrDigit(c)) {
                st.push(String.valueOf(c));
            }
            // If operator
            else {
                // Invalid prefix expression
                if (st.size() < 2) {
                    return "";
                }

                // Pop operands
                String op1 = st.pop(); // first operand
                String op2 = st.pop(); // second operand

                // Form infix expression with parentheses
                String expr = "(" + op1 + c + op2 + ")";

                // Push result back to stack
                st.push(expr);
            }
        }

        // Final validation
        if (st.size() != 1) {
            return "";
        }

        return st.peek();
    }

    // Demo
    public static void main(String[] args) {
        String prefix = "*+AB-C";
        System.out.println(prefixToInfix(prefix)); // (A+B)*(C-?)
    }

    /*
    ------------------------------------------------------------
    TIME COMPLEXITY:
    ------------------------------------------------------------
    O(N)

    ------------------------------------------------------------
    SPACE COMPLEXITY:
    ------------------------------------------------------------
    O(N)
    */
}
